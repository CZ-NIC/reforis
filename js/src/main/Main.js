/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import {BrowserRouter} from 'react-router-dom';
import {Redirect, Switch} from 'react-router';
import PropTypes from 'prop-types';

import {REFORIS_URL_PREFIX} from 'common/constants';
import Portal from 'utils/Portal';
import RouterStateHandler from 'routerStateHandler/RouterStateHandler';
import Navigation from 'navigation/Navigation';

import TopBar from './TopBar';
import {RouteWithSubRoutes} from './routing';
import getPages from './pages';
import {REDIRECT_404_PAGE} from './constants';

Main.propTypes = {
    ws: PropTypes.object.isRequired
};

export default function Main({ws}) {
    const pages = getPages();
    return <BrowserRouter basename={REFORIS_URL_PREFIX}>
        <Portal containerId='navigation_container'>
            <Navigation pages={pages}/>
        </Portal>
        <Portal containerId='topbar_container'>
            <TopBar ws={ws}/>
        </Portal>
        <Portal containerId='router_state_handler_container'>
            <RouterStateHandler ws={ws}/>
        </Portal>

        <Switch>
            {pages.map((route, i) =>
                <RouteWithSubRoutes key={i} ws={ws} {...route}/>
            )}
            <Redirect to={REDIRECT_404_PAGE}/>
        </Switch>
    </BrowserRouter>
}
