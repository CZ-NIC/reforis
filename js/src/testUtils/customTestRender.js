/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from 'react';
import {UIDReset} from 'react-uid';
import {StaticRouter} from 'react-router';
import {render} from '@testing-library/react'

function Wrapper({children}) {
    return <StaticRouter>
        <UIDReset>
            {children}
        </UIDReset>
    </StaticRouter>
}

const customTestRender = (ui, options) => render(ui, {wrapper: Wrapper, ...options});

// re-export everything
export * from '@testing-library/react'

// override render method
export {customTestRender as render}
