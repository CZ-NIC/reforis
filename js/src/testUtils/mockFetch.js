/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

export default function mockFetch(data) {
    return jest.fn(() =>
        new Promise((resolve, reject) => {
            resolve({
                ok: true,
                json: () => {
                    return data ? data : {};
                },
            });
        })
    );
}
