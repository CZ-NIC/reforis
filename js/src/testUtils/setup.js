/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import mockAxios from "jest-mock-axios";
import moment from "moment-timezone";

// Setup axios cleanup
global.afterEach(() => {
    mockAxios.reset();
});

jest.doMock('moment', () => {
    moment.tz.setDefault('UTC');
    return moment;
});

jest.doMock("utils/vfs_fonts", () => ({}));

window.WebSocket = jest.fn();
