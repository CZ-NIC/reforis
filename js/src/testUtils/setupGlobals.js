/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

// Fake babel (gettext) used for docs
global._ = str => str;
global.babel = {format: (str) => str};
global.ForisTranslations = {};
