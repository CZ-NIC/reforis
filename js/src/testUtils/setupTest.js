/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import mockAxios from 'jest-mock-axios';
import moment from 'moment-timezone';

// Setup axios cleanup
global.afterEach(() => {
    mockAxios.reset();
});

// Mock babel (gettext)
window._ = str => str;
window.babel = {format: (str) => str};
window.ForisTranslations = {};

// Mock scrollIntoView
window.HTMLElement.prototype.scrollIntoView = () => {
};

jest.doMock('moment', () => {
    moment.tz.setDefault('UTC');
    return moment;
});

Date.now = jest.fn(() => new Date(Date.UTC(2019, 1, 1, 12, 13, 14)).valueOf());

jest.doMock('utils/vfs_fonts', () => {
    return {};
});
