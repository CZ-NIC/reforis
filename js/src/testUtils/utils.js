/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */


export function createPortalContainer(id) {
    let portalRoot = document.getElementById(id);
        if (!portalRoot) {
            portalRoot = document.createElement('div');
            portalRoot.setAttribute('id', id);
            document.body.appendChild(portalRoot)
        }
}
