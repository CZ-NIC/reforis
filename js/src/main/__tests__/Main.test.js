/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import { render } from "foris/testUtils/customTestRender";
import { WebSockets } from "foris";

import Main from "../Main";

function mockBrokenComponent() {
    // Trigger error
    qwe = asd;
    return <p>Notifications!</p>
}

jest.mock("../pages", () => {
    return () => [
        {
            name: "Notifications",
            path: "/notifications",
            icon: "bell",
            component: mockBrokenComponent,
        },
    ];
});

describe("<Main/>", () => {
    it("should use error boundary when cannot render component", () => {
        global.ForisPlugins = [];
        const webSockets = new WebSockets();
        const { container } = render(
            <>
                <div id="content-container" />
                <Main ws={webSockets} />
            </>,
        );
        expect(container).toMatchSnapshot();
    });
});
