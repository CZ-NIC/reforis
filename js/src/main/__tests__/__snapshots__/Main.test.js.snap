// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main/> should use error boundary when cannot render component 1`] = `
<div>
  <div
    id="content-container"
  >
    <h1>
      An Error Occurred
    </h1>
    <code
      class="error-boundary-description p-3 my-3"
    >
      ReferenceError: asd is not defined
    </code>
    <p>
      More detailed information is available in the console of your web browser - on most browsers accessible after pressing Ctrl+Shift+J or F12.
    </p>
    <p>
      Please report this error to our support team via e-mail: 
      <a
        href="mailto:tech.support@turris.cz"
      >
        tech.support@turris.cz
      </a>
      .
    </p>
  </div>
</div>
`;
